File: python/kswitch/audit/emitter.py
"""
Local audit emitter — writes governed decision events to local JSONL file
and optionally to the server's audit queue.

Audit file: ~/.kswitch/audit/events.jsonl

Event types:
  - enforcement.allow   — local or server ALLOW
  - enforcement.deny    — local or server DENY
  - enforcement.conditional — escalated to server
  - enforcement.revocation_deny — agent in revocation cache

Each event row includes:
  - decision_id, trace_id, bundle_version, context_pack_id
  - policy_set_hash (not available locally — uses bundle_version as proxy)
  - subject tuple (agent_id, mcp_server_id, tool_name)
  - obligations applied, output_policy_mode
  - runtime_mode: LOCAL_RUNTIME_PYTHON
  - timestamp
"""
import json
import os
import threading
import time
import uuid
from typing import Optional, Any

_DEFAULT_AUDIT_DIR = os.path.expanduser("~/.kswitch/audit")
_AUDIT_FILE = "events.jsonl"
_MAX_FILE_SIZE = 50 * 1024 * 1024  # 50 MB — rotate after this


class AuditEmitter:
    """Thread-safe local JSONL audit emitter."""

    def __init__(self, audit_dir: str = _DEFAULT_AUDIT_DIR):
        self._dir = audit_dir
        self._lock = threading.Lock()
        self._sender = None  # Optional central AuditSender (PR-12)

    @property
    def _path(self) -> str:
        return os.path.join(self._dir, _AUDIT_FILE)

    def set_sender(self, sender) -> None:
        """Register a central audit sender. Called by KSwitchRuntime when configured."""
        with self._lock:
            self._sender = sender

    def emit(self, event: dict) -> None:
        """Write one event to the JSONL file and optionally forward to server.

        Step 1 (JSONL) is the first and always-present write.
        Step 2 (central forwarding) is optional and non-blocking.
        A failure in Step 2 never affects Step 1.
        """
        # JSONL is the first and always-present write.
        # Central forwarding is optional and non-blocking.
        # A failure in Step 2 never affects Step 1.

        # Step 1: JSONL write (always, never skipped)
        try:
            os.makedirs(self._dir, exist_ok=True)
            # Rotate if too large
            path = self._path
            if os.path.exists(path) and os.path.getsize(path) > _MAX_FILE_SIZE:
                rotated = path + f".{int(time.time())}"
                try:
                    os.rename(path, rotated)
                except OSError:
                    pass
            with self._lock:
                with open(path, "a", encoding="utf-8") as f:
                    f.write(json.dumps(event, default=str) + "\n")
        except Exception:
            pass  # Audit failure must never fail the governed call

        # Step 2: Central forwarding (best-effort, never blocks)
        try:
            sender = getattr(self, '_sender', None)
            if sender is not None:
                sender.enqueue(event)
        except Exception:
            pass


def _build_event(
    event_type: str,
    agent_id: str,
    mcp_server_id: str,
    tool_name: str,
    allowed: bool,
    reason: str,
    decision_id: str,
    decision_path: list,
    obligations: list,
    output_policy: Optional[dict],
    evaluation_mode: str,
    bundle_version: str = "",
    context_pack_id: str = "",
    risk_tier: str = "medium",
    elapsed_ms: float = 0.0,
) -> dict:
    """Build a structured audit event dict."""
    outcome = "allow" if allowed else "deny"
    return {
        "event_id": str(uuid.uuid4()),
        "event_type": event_type,
        "event_version": "1.0",

        # Subject
        "agent_id": agent_id,
        "mcp_server_id": mcp_server_id,
        "tool_name": tool_name or "",
        "action": "mcp_call",

        # Decision
        "decision_id": decision_id,
        "allowed": allowed,
        "outcome": outcome,
        "reason": reason,
        "decision_path": decision_path,

        # Obligations
        "obligations": obligations or [],
        "output_policy_mode": (output_policy or {}).get("mode", ""),

        # Provenance
        "bundle_version": bundle_version,
        "context_pack_id": context_pack_id,
        "risk_tier": risk_tier,
        "runtime_mode": evaluation_mode,

        # Timing
        "elapsed_ms": elapsed_ms,
        "evaluated_at": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
    }


# Module-level singleton
_emitter = AuditEmitter()


def emit_decision_event(
    event_type: str,
    agent_id: str,
    mcp_server_id: str,
    tool_name: str,
    allowed: bool,
    reason: str,
    decision_id: str = "",
    decision_path: list = None,
    obligations: list = None,
    output_policy: Optional[dict] = None,
    evaluation_mode: str = "LOCAL_RUNTIME_PYTHON",
    bundle_version: str = "",
    context_pack_id: str = "",
    risk_tier: str = "medium",
    elapsed_ms: float = 0.0,
) -> None:
    event = _build_event(
        event_type=event_type,
        agent_id=agent_id,
        mcp_server_id=mcp_server_id,
        tool_name=tool_name,
        allowed=allowed,
        reason=reason,
        decision_id=decision_id or str(uuid.uuid4()),
        decision_path=decision_path or [],
        obligations=obligations or [],
        output_policy=output_policy,
        evaluation_mode=evaluation_mode,
        bundle_version=bundle_version,
        context_pack_id=context_pack_id,
        risk_tier=risk_tier,
        elapsed_ms=elapsed_ms,
    )
    _emitter.emit(event)


def get_audit_emitter() -> AuditEmitter:
    return _emitter

File: go/kswitch/audit/emitter.go
// Package audit provides local audit event emission for the Go SDK.
//
// Audit file: {stateDir}/audit/events.jsonl
//
// Event schema matches Python AuditEmitter and TypeScript AuditEmitter exactly:
//   - event_id, event_type, event_version
//   - agent_id, mcp_server_id, tool_name, action
//   - decision_id, allowed, outcome, reason, decision_path
//   - obligations, output_policy_mode
//   - bundle_version, context_pack_id, risk_tier, runtime_mode ("LOCAL_RUNTIME_GO")
//   - elapsed_ms, evaluated_at
//
// Write order (same as Python/TypeScript):
//   Step 1: JSONL write to local file — always, never skipped.
//   Step 2: Enqueue to AuditSender for central forwarding — optional, non-blocking.
//   A failure in Step 2 never affects Step 1 or the decision path.
package audit

import (
	"encoding/json"
	"fmt"
	"log/slog"
	"os"
	"path/filepath"
	"sync"
	"time"

	"github.com/google/uuid"
)

const (
	auditFileName = "events.jsonl"
	maxFileSize   = 50 * 1024 * 1024 // 50 MB — rotate after this
	runtimeMode   = "LOCAL_RUNTIME_GO"
)

// AuditEvent is the structured event emitted for every governance decision.
type AuditEvent struct {
	EventID      string   `json:"event_id"`
	EventType    string   `json:"event_type"`
	EventVersion string   `json:"event_version"`
	AgentID      string   `json:"agent_id"`
	MCPServerID  string   `json:"mcp_server_id"`
	ToolName     string   `json:"tool_name"`
	Action       string   `json:"action"`
	DecisionID   string   `json:"decision_id"`
	Allowed      bool     `json:"allowed"`
	Outcome      string   `json:"outcome"`
	Reason       string   `json:"reason"`
	DecisionPath []string `json:"decision_path"`
	Obligations  []any    `json:"obligations"`
	OutputPolicyMode string `json:"output_policy_mode"`
	BundleVersion  string `json:"bundle_version"`
	ContextPackID  string `json:"context_pack_id"`
	RiskTier       string `json:"risk_tier"`
	RuntimeMode    string `json:"runtime_mode"`
	ElapsedMs      float64 `json:"elapsed_ms"`
	EvaluatedAt    string  `json:"evaluated_at"`
}

// BuildAuditEvent constructs an AuditEvent from individual fields.
// Mirrors Python _build_event() and TypeScript buildAuditEvent().
func BuildAuditEvent(
	eventType, agentID, mcpServerID, toolName string,
	allowed bool,
	reason, decisionID string,
	decisionPath []string,
	obligations []any,
	outputPolicyMode string,
	bundleVersion, contextPackID, riskTier string,
	elapsedMs float64,
) AuditEvent {
	outcome := "deny"
	if allowed {
		outcome = "allow"
	}
	if decisionID == "" {
		decisionID = uuid.New().String()
	}
	if decisionPath == nil {
		decisionPath = []string{}
	}
	if obligations == nil {
		obligations = []any{}
	}
	return AuditEvent{
		EventID:          uuid.New().String(),
		EventType:        eventType,
		EventVersion:     "1.0",
		AgentID:          agentID,
		MCPServerID:      mcpServerID,
		ToolName:         toolName,
		Action:           "mcp_call",
		DecisionID:       decisionID,
		Allowed:          allowed,
		Outcome:          outcome,
		Reason:           reason,
		DecisionPath:     decisionPath,
		Obligations:      obligations,
		OutputPolicyMode: outputPolicyMode,
		BundleVersion:    bundleVersion,
		ContextPackID:    contextPackID,
		RiskTier:         riskTier,
		RuntimeMode:      runtimeMode,
		ElapsedMs:        elapsedMs,
		EvaluatedAt:      time.Now().UTC().Format(time.RFC3339),
	}
}

// Sender is the interface the AuditEmitter uses to forward events centrally.
type Sender interface {
	Enqueue(event AuditEvent)
}

// AuditEmitter writes governance decision events to a local JSONL file and
// optionally forwards them to the central server. Safe for concurrent use.
type AuditEmitter struct {
	dir    string
	mu     sync.Mutex
	sender Sender // optional — nil if central forwarding not configured
}

// NewAuditEmitter creates an emitter writing to dir.
// If dir is empty, DefaultAuditDir() is used.
func NewAuditEmitter(dir string) *AuditEmitter {
	if dir == "" {
		dir = DefaultAuditDir()
	}
	return &AuditEmitter{dir: dir}
}

// DefaultAuditDir returns the default audit log directory.
// Resolution order:
//  1. $KSWITCH_STATE_DIR/audit
//  2. $HOME/.kswitch/audit
func DefaultAuditDir() string {
	if s := os.Getenv("KSWITCH_STATE_DIR"); s != "" {
		return filepath.Join(s, "audit")
	}
	home, err := os.UserHomeDir()
	if err != nil {
		return ""
	}
	return filepath.Join(home, ".kswitch", "audit")
}

func (e *AuditEmitter) filePath() string {
	return filepath.Join(e.dir, auditFileName)
}

// SetSender registers a central audit sender.
// Called by the runtime when central forwarding is configured.
func (e *AuditEmitter) SetSender(s Sender) {
	e.mu.Lock()
	e.sender = s
	e.mu.Unlock()
}

// Emit writes one event to the local JSONL file and optionally forwards it centrally.
// Never returns an error — audit failures must never block the decision path.
func (e *AuditEmitter) Emit(event AuditEvent) {
	// ── Step 1: JSONL write (always, never skipped) ──────────────────────────
	func() {
		defer func() {
			if r := recover(); r != nil {
				slog.Debug("kswitch.audit: JSONL write panic", "error", r)
			}
		}()
		if e.dir == "" {
			return
		}
		if err := os.MkdirAll(e.dir, 0o755); err != nil {
			return
		}
		path := e.filePath()

		// Rotate if too large.
		if info, err := os.Stat(path); err == nil && info.Size() > maxFileSize {
			rotated := fmt.Sprintf("%s.%d", path, time.Now().Unix())
			_ = os.Rename(path, rotated)
		}

		line, err := json.Marshal(event)
		if err != nil {
			return
		}
		e.mu.Lock()
		f, err := os.OpenFile(path, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o644)
		if err != nil {
			e.mu.Unlock()
			return
		}
		_, _ = f.Write(append(line, '\n'))
		_ = f.Close()
		e.mu.Unlock()
	}()

	// ── Step 2: Central forwarding (best-effort, never blocks) ───────────────
	func() {
		defer func() {
			if r := recover(); r != nil {
				slog.Debug("kswitch.audit: sender enqueue panic", "error", r)
			}
		}()
		e.mu.Lock()
		sender := e.sender
		e.mu.Unlock()
		if sender == nil {
			return
		}
		sender.Enqueue(event)
	}()
}

// ── Module-level singleton ────────────────────────────────────────────────────

var (
	defaultEmitter     *AuditEmitter
	defaultEmitterOnce sync.Once
)

func getDefaultEmitter() *AuditEmitter {
	defaultEmitterOnce.Do(func() {
		defaultEmitter = NewAuditEmitter("")
	})
	return defaultEmitter
}

// EmitDecisionEvent emits a governance decision event using the default emitter.
func EmitDecisionEvent(
	eventType, agentID, mcpServerID, toolName string,
	allowed bool,
	reason, decisionID string,
	decisionPath []string,
	obligations []any,
	outputPolicyMode string,
	bundleVersion, contextPackID, riskTier string,
	elapsedMs float64,
) {
	event := BuildAuditEvent(
		eventType, agentID, mcpServerID, toolName,
		allowed, reason, decisionID,
		decisionPath, obligations, outputPolicyMode,
		bundleVersion, contextPackID, riskTier, elapsedMs,
	)
	getDefaultEmitter().Emit(event)
}

// GetAuditEmitter returns the module-level default emitter.
func GetAuditEmitter() *AuditEmitter {
	return getDefaultEmitter()
}

// SetAuditEmitter replaces the module-level default emitter (for testing).
func SetAuditEmitter(em *AuditEmitter) {
	defaultEmitterOnce = sync.Once{}
	defaultEmitterOnce.Do(func() { defaultEmitter = em })
}
